hal-core/src/mem/page.rs

```diff
@@ -0,0 +1,204 @@
+    ///
+    /// Note that an implementation of this method is provided as long as an
+    /// implementor of this trait provides `dealloc_range`.
+    ///
+    /// # Returns
+    /// - `Ok(())` if the page was successfully deallocated.
+    /// - `Err` if the requested range could not be deallocated.
+    fn dealloc(&mut self, page: Page<A::PAddr, S>) -> Result<(), AllocErr> {
+        self.dealloc_range(page.range_inclusive(page))
+    }
+
+    /// Deallocate a range of pages.
+    ///
+    /// # Returns
+    /// - `Ok(())` if a range of pages was successfully deallocated
+    /// - `Err` if the requested range could not be deallocated.
+    fn dealloc_range(&self, range: PageRange<A::PAddr, S>) -> Result<(), AllocErr>;
+}
+
+/// A memory page.
+#[derive(Clone, Copy, PartialEq, Eq)]
+#[repr(C)]
+pub struct Page<A, S: Size> {
+    base: A,
+    _size: PhantomData<S>,
+}
+
+/// A range of memory pages of the same size.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct PageRange<A: Address, S: Size> {
+    start: Page<A, S>,
+    end: Page<A, S>,
+}
+
+pub struct NotAligned<S> {
+    _size: PhantomData<S>,
+}
+
+pub struct AllocErr {
+    // TODO: eliza
+    _p: (),
+}
+
+// === impl Page ===
+
+impl<A: Address, S: Size> Page<A, S> {
+    /// Returns a page starting at the given address.
+    pub fn starting_at(addr: impl Into<A>) -> Result<Self, NotAligned<S>> {
+        let addr = addr.into();
+        if !addr.is_aligned(S::SIZE) {
+            return Err(NotAligned { _size: PhantomData });
+        }
+        Ok(Self::containing(addr))
+    }
+
+    /// Returns the page that contains the given address.
+    pub fn containing(addr: impl Into<A>) -> Self {
+        let base = addr.into().align_down(S::SIZE);
+        Self {
+            base,
+            _size: PhantomData,
+        }
+    }
+
+    pub fn base_address(&self) -> A {
+        self.base
+    }
+
+    pub fn end_address(&self) -> A {
+        self.base + S::SIZE
+    }
+
+    pub fn contains(&self, addr: impl Into<A>) -> bool {
+        let addr = addr.into();
+        addr >= self.base && addr <= self.end_address()
+    }
+
+    pub fn range_inclusive(self, end: Page<A, S>) -> PageRange<A, S> {
+        PageRange { start: self, end }
+    }
+
+    pub fn range_to(self, end: Page<A, S>) -> PageRange<A, S> {
+        PageRange {
+            start: self,
+            end: end - 1,
+        }
+    }
+}
+
+impl<A: Address, S: Size> ops::Add<usize> for Page<A, S> {
+    type Output = Self;
+    fn add(self, rhs: usize) -> Self {
+        Page {
+            base: self.base + (S::SIZE * rhs),
+            _size: PhantomData,
+        }
+    }
+}
+
+impl<A: Address, S: Size> ops::Sub<usize> for Page<A, S> {
+    type Output = Self;
+    fn sub(self, rhs: usize) -> Self {
+        Page {
+            base: self.base - (S::SIZE * rhs),
+            _size: PhantomData,
+        }
+    }
+}
+
+impl<A: Address, S: Size> AsRef<[u8]> for Page<A, S> {
```

Comment:
i was going to ask if there should be an `AsMut` too, but if we're only reading pages to like, serialize them or something, and we write to pages by "writing to pointers" and letting the mmu deal with it, we don't need `AsMut` huh?

Reply:
upon further consideration i think this should probably be a method rather than an AsRef impl, so that it can be unsafe to call...you would need to ensure nothing is concurrently writing to the page.

Reply:
haha, supporting concurrency.... (this is a very good point)